include/ocpp/v201/functional_blocks/authorization.hpp

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Pionix GmbH and Contributors to EVerest
+
+#pragma once
+
+#include <ocpp/v201/messages/Authorize.hpp>
+#include <ocpp/v201/messages/ClearCache.hpp>
+#include <ocpp/v201/messages/GetLocalListVersion.hpp>
+#include <ocpp/v201/messages/SendLocalList.hpp>
+
+#include <ocpp/v201/database_handler.hpp>
+#include <ocpp/v201/message_dispatcher.hpp>
+#include <ocpp/v201/message_handler.hpp>
+
+namespace ocpp::v201 {
+class AuthorizationInterface : public MessageHandlerInterface {
+public:
+    virtual ~AuthorizationInterface() {
+    }
+
+    virtual void start() = 0;
+    virtual AuthorizeResponse authorize_req(const IdToken id_token, const std::optional<CiString<5500>>& certificate,
+                                            const std::optional<std::vector<OCSPRequestData>>& ocsp_request_data) = 0;
+    virtual void trigger_authorization_cache_cleanup() = 0;
+    ///\brief Calculate and update the authorization cache size in the device model
+    ///
+    virtual void update_authorization_cache_size() = 0;
+    virtual bool is_auth_cache_ctrlr_enabled() = 0;
+    virtual void authorization_cache_insert_entry(const std::string& id_token_hash,
+                                                  const IdTokenInfo& id_token_info) = 0;
+    virtual std::optional<AuthorizationCacheEntry> authorization_cache_get_entry(const std::string& id_token_hash) = 0;
+    virtual void authorization_cache_delete_entry(const std::string& id_token_hash) = 0;
+
+    /// \brief Validates provided \p id_token \p certificate and \p ocsp_request_data using CSMS, AuthCache or AuthList
+    /// \param id_token
+    /// \param certificate
+    /// \param ocsp_request_data
+    /// \return AuthorizeResponse containing the result of the validation
+    virtual AuthorizeResponse validate_token(const IdToken id_token, const std::optional<CiString<5500>>& certificate,
+                                             const std::optional<std::vector<OCSPRequestData>>& ocsp_request_data) = 0;
+};
+
+class Authorization : public AuthorizationInterface {
+private: // Members
+    MessageDispatcherInterface<MessageType>& message_dispatcher;
+    DeviceModel& device_model;
+    ConnectivityManagerInterface& connectivity_manager;
+    std::shared_ptr<ocpp::v201::DatabaseHandlerInterface> database_handler;
+    std::shared_ptr<EvseSecurity> evse_security;
```

Comment:
Is there a reason why these members are not defined as references like the other ones?

Reply:
In the interest of slowly moving away from having everything shared, and instead have explicit ownership it might be good to already use references instead. Then we can change it to unique_ptr later